db/blob/blob_source.cc

```diff
@@ -0,0 +1,212 @@
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/blob/blob_source.h"
+
+#include <cassert>
+#include <string>
+
+#include "db/blob/blob_file_reader.h"
+#include "file/file_prefetch_buffer.h"
+#include "options/cf_options.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+BlobSource::BlobSource(Cache* cache, const ImmutableOptions* immutable_options,
+                       const FileOptions* file_options,
+                       const std::string& db_id,
+                       const std::string& db_session_id,
+                       uint32_t column_family_id,
+                       HistogramImpl* blob_file_read_hist,
+                       const std::shared_ptr<IOTracer>& io_tracer)
+    : db_id_(db_id),
+      db_session_id_(db_session_id),
+      statistics_(immutable_options->statistics.get()),
+      blob_file_cache_(new BlobFileCache(cache, immutable_options, file_options,
+                                         column_family_id, blob_file_read_hist,
+                                         io_tracer)),
+      blob_cache_(immutable_options->blob_cache) {}
+
+BlobSource::~BlobSource() = default;
+
+Status BlobSource::MaybeReadBlobFromCache(
+    FilePrefetchBuffer* prefetch_buffer, const ReadOptions& read_options,
+    const CacheKey& cache_key, uint64_t offset,
+    CachableEntry<Slice>* blob_entry) const {
+  assert(blob_entry != nullptr);
+  assert(blob_entry->IsEmpty());
+
+  // First, try to get the blob from the cache
+  //
+  // If blob cache is enabled, we'll try to read from it.
+  Status s;
+
+  if (blob_cache_ != nullptr) {
+    Slice key = cache_key.AsSlice();
+    s = GetBlobFromCache(key, blob_cache_.get(), blob_entry);
+    if (blob_entry->GetValue()) {
+      if (prefetch_buffer) {
+        // Update the blob details so that PrefetchBuffer can use the read
+        // pattern to determine if reads are sequential or not for prefetching.
+        // It should also take in account blob read from cache.
+        prefetch_buffer->UpdateReadPattern(
+            offset, blob_entry->GetValue()->size(),
+            read_options.adaptive_readahead /* decrease_readahead_size */);
+      }
+    }
+  }
+
+  assert(s.ok() || blob_entry->GetValue() == nullptr);
+
+  return s;
+}
+
+Status BlobSource::GetBlobFromCache(const Slice& cache_key, Cache* blob_cache,
+                                    CachableEntry<Slice>* blob) const {
+  assert(blob);
+  assert(blob->IsEmpty());
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  if (blob_cache != nullptr) {
+    assert(!cache_key.empty());
+    Cache::Handle* cache_handle = nullptr;
+    cache_handle = GetEntryFromCache(blob_cache, cache_key, priority);
+    if (cache_handle != nullptr) {
+      blob->SetCachedValue(
+          reinterpret_cast<Slice*>(blob_cache->Value(cache_handle)), blob_cache,
```

Comment:
You're probably copying some existing code, but static_cast is generally preferred where applicable. https://stackoverflow.com/questions/310451/should-i-use-static-cast-or-reinterpret-cast-when-casting-a-void-to-whatever

Reply:
Thanks, @pdillinger !